### DISCUSSION TITLE
It seems that variable substitution is not properly working

### DISCUSSION THREAD

**Discussion by elmuz** [fail2ban/fail2ban#3865] [Q&A] [answered]
### Summary
Maybe I haven't fully understood the general idea of the configuration files. However, here's my issue. I am using the  'cloudflare-token' action, but it seems to me that internal variables like for example `<cfzone>` is not substituted when executing the ban/unban command. Instead it is passed literally as `<cfzone>`.

I believe I did some mistake in the configuration, but I've checked many times and I can't find any. These are the steps to get into my situation.

### Steps to reproduce
1. Create a file `action.d/cloudflare-token.local` with toy example of `actionban` (in my case `echo "<cfzone>" > /config/fail2ban/test.txt`
2. Set variable `cfzone = ...` in the `jail.local`
3. trigger the ban action
4. check the file

### Other details
- I am using fail2ban as part of the swag reverse-proxy container from linuxserver.io.
- Other variables from `jail.local` are properly interpreted (like `maxretries` or `bantime`).
-  I am using jails for Authelia in this case.
- As mentioned in title I am using 'cloudflare tunnel'.
- Ban action is properly triggered so configuration of the jail and the filter for the log file are correct.

I feel there's something maybe related with angular brackets escaping, or single/double quotes.
Any idea?

**Comment by sebres** [comment, marked as answer]
The variables set in jail section wouldn't necessarily interpolated in the action directly. They have basically different substitution scopes, unless they would also supplied to the action as parameters or they belong to the standard ticket-related tags.
With other words, one has to supply them to the action as parameters, see for instance many definitions and examples in the jail.conf (e.g. how port, protocol etc will be supplied):
https://github.com/fail2ban/fail2ban/blob/0bf1106d72094b95996fa9b75ec76235d64703d3/config/jail.conf#L211-L221

And other values, you mean to be correctly interpreted, are special tags representing the ticket info and not the configuration (and they will be substituted directly):
https://github.com/fail2ban/fail2ban/blob/0bf1106d72094b95996fa9b75ec76235d64703d3/fail2ban/server/actions.py#L374-L401

Also note https://github.com/fail2ban/fail2ban/wiki/How-fail2ban-substitution-resp.-runtime-interpolation-works

**Comment by elmuz** [comment]
oh god was already in the wiki. Sorry about that. Clear now, thank you.
